src/scene/shader-lib/programs/lit-shader.js

```diff
@@ -1244,22 +1210,63 @@ class LitShader {
                     }
 
                     if (shadowReadMode !== null) {
+                        if (light._normalOffsetBias) {
+                            func.append("#define SHADOW_SAMPLE_NORMAL_OFFSET");
+                        }
+                        if (lightType === LIGHTTYPE_DIRECTIONAL) {
+                            func.append("#define SHADOW_SAMPLE_ORTHO");
+                        }
+                        if ((light._isPcf) &&
+                            device.webgl2 || device.extStandardDerivatives || device.isWebGPU) {
+                            func.append("#define SHADOW_SAMPLE_SOURCE_ZBUFFER");
+                        }
                         if (lightType === LIGHTTYPE_OMNI) {
-                            const shadowCoordArgs = "(light" + i + "_shadowMap, dShadowCoord, light" + i + "_shadowParams, dLightDirW);";
-                            if (light._normalOffsetBias) {
-                                backend.append("    normalOffsetPointShadow(light" + i + "_shadowParams, dLightPosW, dLightDirW, dLightDirNormW, dVertexNormalW);");
-                            }
-                            backend.append(`    float shadow${i} = getShadowPoint${shadowReadMode}${shadowCoordArgs}`);
-                            backend.append(`    dAtten *= mix(1.0, shadow${i}, light${i}_shadowIntensity);`);
-                        } else {
-                            const shadowMatArg = `light${i}_shadowMatrix`;
-                            const shadowParamArg = `light${i}_shadowParams`;
-                            backend.append(this._nonPointShadowMapProjection(device, options.lights[i], shadowMatArg, shadowParamArg, i));
+                            func.append("#define SHADOW_SAMPLE_POINT");
+                        }
+
+                        // Create shadow coord sampler function for this light
+                        const coordCode = chunks.shadowSampleCoordPS;
+                        func.append(coordCode.replace("$LIGHT", i));
+
+                        func.append("#undef SHADOW_SAMPLE_NORMAL_OFFSET");
+                        func.append("#undef SHADOW_SAMPLE_ORTHO");
+                        func.append("#undef SHADOW_SAMPLE_SOURCE_ZBUFFER");
+                        func.append("#undef SHADOW_SAMPLE_POINT");
+
+                        // Make sure to undefine the shadow sampler defines
```

Comment:
should this line be 4 lines higher, before those undefs?

Reply:
The code is basically, set defines, generate chunk with defines set, unset the defines - for each light.

Reply:
sorry, I just meant to move the comment above those lines that do the under.